### PR TITLE
fix: support for OpenResty 1.11.2.*

### PR DESCRIPTION
**Pull request by thibaultcha**
This PR cleans up and replaces #16

### CODE REVIEW

File: kong-ngx-build [hunk 1]
```diff
@@ -329,11 +338,17 @@ main() {
...
         )
 
+        if version_gte $OPENRESTY_VER 1.11.4; then
+          OPENRESTY_OPTS+=('--with-stream_realip_module')
+        fi
+
+        if version_gte $OPENRESTY_VER 1.11.5; then
+          OPENRESTY_OPTS+=('--with-stream_ssl_preread_module')
+        fi
+
```
Comment (eskerda) [inline kong-ngx-build]: It is equivalent, but should these checks be against `$NGINX_CORE_VER` instead?

File: kong-ngx-build [hunk 2]
```diff
@@ -387,6 +402,81 @@ main() {
   succ "Build finished in $SECONDS seconds. Enjoy!"
 }
 
+parse_version() {
+  [[ -z $1 ]] && fatal 'missing arg $1 when invoking parse_version()'
+  [[ -z $2 ]] && fatal 'missing arg $2 when invoking parse_version()'
+
+  # remove non-version parts of a version, ie foobar-1.2.3a.5-rc1 -> 1.2.3a.5
```
Comment (eskerda) [inline kong-ngx-build]: We should update this comment saying it only removes the prefix on versions! Which I think is ok, if we are not going to support dash releases

File: kong-ngx-build (hunk 1 above)
Comment (thibaultcha) [reply kong-ngx-build]: @eskerda They sure can! I will reorder the commits and slightly change the fix to use it :+1:

File: kong-ngx-build [hunk 3]
```diff
@@ -164,7 +166,6 @@ main() {
         echo "$OPENSSL_SHA openssl-$OPENSSL_VER.tar.gz" | sha256sum -c -
       fi
       tar -xzvf openssl-$OPENSSL_VER.tar.gz
-      ln -s openssl-$OPENSSL_VER openssl
```
Comment (hutchic) [inline kong-ngx-build]: these symlinks were at one point used by downstream scripts which were unaware of the version of openssl / pcre we're using. I think that's no longer the case. For my own knowledge what was the issue with having them?

File: .travis.yml [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+install:
+  - if [[ $NO_OPENRESTY_PATCHES == "1" ]]; then OPENRESTY_PATCHES=--no-openresty-patches; fi
+  - echo $OPENRESTY_PATCHES
+
+script:
+  - ./kong-ngx-build --work $WORK --prefix $PREFIX --jobs $JOBS --openresty $OPENRESTY --openssl $OPENSSL --luarocks $LUAROCKS --pcre $PCRE $OPENRESTY_PATCHES >build.log 2>&1 || (cat build.log; exit 1)
+  - export PATH=$PREFIX/openresty/bin:$PATH
+  - export PATH=$PREFIX/luarocks/bin:$PATH
+  - export PATH=$PREFIX/openssl/bin:$PATH
```
Comment (hutchic) [inline .travis.yml]: could move these `exports` to `env: global:` but not important

File: .travis.yml [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  - (openresty -v 2>&1 | grep -F -- "$OPENRESTY") || (openresty -v; exit 1)
+  - (openssl version | grep -F -- "$OPENSSL") || (openssl version; exit 1)
+  - (luarocks --version | grep -F -- "$LUAROCKS") || (luarocks --version; exit 1)
+  - (ldd $(which openresty) | grep -E -- "libssl.*?$PREFIX") || (ldd $(which openresty); exit 1)
+  - (ldd $(which openresty) | grep -E -- "libcrypto.*?$PREFIX") || (ldd $(which openrest)y; exit 1)
+  - (ldd $(which openresty) | grep -E -- "libluajit.*?$PREFIX") || (ldd $(which openresty); exit 1)
+  - (openresty -V 2>&1 | grep -F -- "--with-pcre=$WORK/pcre-$PCRE") || (openresty -V; exit 1)
+  - if [[ $NO_OPENRESTY_PATCHES == "1" ]]; then [[ ! -f "$WORK/openresty-$OPENRESTY/bundle/.patch_applied" ]] || (echo "OpenResty patches mistakenly applied"; exit 1); fi
+  - if [[ $NO_OPENRESTY_PATCHES != "1" ]]; then [[ -f "$WORK/openresty-$OPENRESTY/bundle/.patch_applied" ]] || (echo "OpenResty patches not applied"; exit 1); fi
+  - openresty -V # for humans
```
Comment (hutchic) [inline .travis.yml]: I like these tests A LOT thank you!

Could you add a `luarocks install version` (or any arbitrary package) just to ensure that `luarocks` knows how to locate `luajit`?

File: .travis.yml [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+compiler:
+  - gcc
+  - clang
+
+env:
+  global:
+    - JOBS=2
+    - WORK=$TRAVIS_BUILD_DIR/work
+    - PREFIX=$HOME/kopenresty
+    - LUAROCKS=3.1.3
```
Comment (hutchic) [inline .travis.yml]: should `LUAROCKS` be a `matrix` option instead?

File: kong-ngx-build (hunk 2 above)
Comment (thibaultcha) [reply kong-ngx-build]: I just rewrote the parsing utility, and as a bonus, got rid of the Perl dependency, made the parsing  70% faster, and consider `-rcN` suffixes!

File: kong-ngx-build (hunk 3 above)
Comment (thibaultcha) [reply kong-ngx-build]: 1. If we ask for a different version of one of those dependencies in a subsequent run, and both versions are already downloaded, the symlink will not be pointing to the right version.
2. The PCRE symlink was invalid (pointing to a non-existing directory)
3. If we don't need them, let's get rid of them :)

File: .travis.yml (hunk 6 above)
Comment (thibaultcha) [reply .travis.yml]: Don't think it's worth it. It'd significantly increase the build time, it isn't tightly coupled (in terms of dependency requirements) with either of OpenResty/OpenSSL, and is already part of CI in other repositories, including Kong/kong.

File: .travis.yml (hunk 5 above)
Comment (thibaultcha) [reply .travis.yml]: If it couldn't locate LuaJIT, `luarocks --version` would already fail (or would have not been installed in the first place).

Comment (thibaultcha) [reply .travis.yml]: Seems like when encountering the above, the command still exists with `0`. I added such cases (although without depending on installing modules, which feels like too much and too slow, as well as too unreliable - we've had spurious failures due to connectivity issues with LuaRocks/GitHub in our CI before).

Comment (thibaultcha) [reply .travis.yml]: Those new tests just caught `DESTDIR` not being very nice:

https://travis-ci.com/Kong/openresty-build-tools/jobs/211662676#L417

So I just pushed yet another fix making `*_DESTDIR` variables more robust and less error-prompt for future maintainers as well.

Comment (hutchic) [reply .travis.yml]: yeah @gszr and I hit this today which is what spurred me to ask for the test 👍

**Comment by hutchic** [comment]
need to make sure we capture the use case of compiling / building to one location with the intention of packaging and installing to another.

Specifically
```
            --with-lua=$OPENRESTY_INSTALL/luajit \
            --with-lua-include=$OPENRESTY_INSTALL/luajit/include/luajit-2.1
```
needs to reference the *final* end result but during compilation / buildling the location also has to exist. I don't like and hope we can find a better solution then the one where I double `make install`'d

**Comment by hutchic** [comment]
Test scenario:
```
LUAROCKS_PREFIX=/usr/local \
LUAROCKS_DESTDIR=/tmp/build \
OPENRESTY_PREFIX=/usr/local/openresty \
OPENRESTY_DESTDIR=/tmp/build \
OPENSSL_PREFIX=/usr/local/kong \
OPENSSL_DESTDIR=/tmp/build \
/tmp/kong-ngx-build -p /tmp/build/usr/local \
--openresty $RESTY_VERSION \
--openssl $RESTY_OPENSSL_VERSION \
--luarocks $RESTY_LUAROCKS_VERSION \
--pcre $RESTY_PCRE_VERSION \
--work /work
cp -R /tmp/build/* /
```
Result:
```
luarocks config

   LUA = "/tmp/build/usr/local/openresty/luajit/bin/luajit",
   LUALIB = "libluajit-5.1.a",
   LUA_BINDIR = "/tmp/build/usr/local/openresty/luajit/bin",
   LUA_DIR = "/tmp/build/usr/local/openresty/luajit",
   LUA_INCDIR = "/tmp/build/usr/local/openresty/luajit/include/luajit-2.1",
   LUA_LIBDIR = "/tmp/build/usr/local/openresty/luajit/lib",
```

**Comment by thibaultcha** [comment]
Well, fixing this use-case is outside the scope of this PR (which already extended quite beyond its original one...) I will be merging this for now :)

File: kong-ngx-build [hunk 7]
```diff
@@ -387,6 +407,83 @@ main() {
   succ "Build finished in $SECONDS seconds. Enjoy!"
 }
 
+parse_version() {
+  [[ -z $1 ]] && fatal 'missing arg $1 when invoking parse_version()'
+  [[ -z $2 ]] && fatal 'missing arg $2 when invoking parse_version()'
+
+  local subj=$1
+  local -n ver=$2
```
Comment (eskerda) [inline kong-ngx-build]: Note `local -n` only works on bash >= 5, which is not that new, but for instance, os x comes bundled with 3.2

File: kong-ngx-build [hunk 8]
```diff
@@ -387,6 +407,83 @@ main() {
...
+
+version_lt() {
+  local version_a version_b
+
+  parse_version $1 version_a
+  parse_version $2 version_b
+
+  for index in "${!version_a[@]}"; do
+    [[ "${version_a[index]}" < "${version_b[index]}" ]] && return 0
+    [[ "${version_a[index]}" > "${version_b[index]}" ]] && return 1
```
Comment (eskerda) [inline kong-ngx-build]: These comparisons do not work on all cases, since these are string comparisons `"99" > "100"`

Comment (thibaultcha) [reply kong-ngx-build]: Ok, let's add more test cases covering those then, thanks for the catch :+1:

File: kong-ngx-build (hunk 7 above)
Comment (thibaultcha) [reply kong-ngx-build]: Worst case, we can revert this commit and go back to the previous implementation, I made sure to not override your first patch.

**Pull request merged**